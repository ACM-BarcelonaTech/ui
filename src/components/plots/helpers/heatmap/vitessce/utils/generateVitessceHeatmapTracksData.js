import getCellClassProperties from 'utils/cellSets/getCellClassProperties';
import { hexToRgb } from 'utils/plotUtils';

const generateVitessceHeatmapTracksData = (trackOrder, cellSets, cells) => {
  const colorForCell = (cellId, cellClassKey) => {
    const { color: cellColor = null } = getCellClassProperties(
      cellId,
      [cellClassKey],
      cellSets,
    )[0] ?? {};
    return hexToRgb(cellColor) ?? hexToRgb('#f5f8fa');
  };

  const cellIdsColorsMap = new Map();

  cells.forEach((cellId) => {
    const allColorsForCell = trackOrder.map((cellClassKey) => colorForCell(cellId, cellClassKey));

    cellIdsColorsMap.set(`${cellId}`, allColorsForCell);
  });

  return cellIdsColorsMap;
};

export default generateVitessceHeatmapTracksData;
