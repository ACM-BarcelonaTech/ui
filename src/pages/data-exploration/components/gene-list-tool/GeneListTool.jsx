import React, { useState } from 'react';
import {
  useSelector, useDispatch,
} from 'react-redux';

import {
  Table, Space,
} from 'antd';

import PropTypes from 'prop-types';
import FilterGenes from './FilterGenes';
import { updateGeneList } from '../../../../redux/actions';

const GeneListTool = (props) => {
  const { experimentID } = props;
  const dispatch = useDispatch();

  const isLoading = useSelector((state) => state.geneList.loading);
  const rows = useSelector((state) => state.geneList.rows);
  const tableState = useSelector((state) => state.geneList.tableState);
  const [geneNamesFilter, setGeneNamesFilter] = useState(null);

  if (!tableState) {
    const defaultState = {
      pagination: {
        current: 1,
        pageSize: 15,
        showSizeChanger: true,
        total: 1,
      },
      sorter: {
        field: 'dispersions',
        order: 'descend',
      },
      geneNamesFilter,
    };

    dispatch(updateGeneList(experimentID, defaultState));
  }

  const getSortOrderIfExists = (key) => {
    if (key === tableState?.sorter.columnKey) {
      return tableState.sorter.order;
    }
    return null;
  };

  const columns = [
    {
      title: 'Gene',
      dataIndex: 'gene_names',
      key: 'gene_names',
      sorter: true,
      render: (geneName) => (
        <a
          href={`https://www.genecards.org/cgi-bin/carddisp.pl?gene=${geneName}`}
          target='_blank'
          rel='noreferrer'
        >
          {geneName}
        </a>
      ),
      sortOrder: getSortOrderIfExists('gene_names'),
    },
    {
      title: 'Dispersion',
      dataIndex: 'dispersions',
      key: 'dispersions',
      sorter: true,
      sortOrder: getSortOrderIfExists('dispersions'),
      render: (num) => parseFloat(num.toFixed(3)),
    },
  ];

  const handleTableChange = (newPagination, _, newSorter) => {
    const newTableState = { pagination: newPagination, sorter: { ...newSorter }, geneNamesFilter };
    dispatch(updateGeneList(experimentID, newTableState));
  };

  const filterGenes = (searchPattern) => {
    const newTableState = { pagination: tableState.pagination, sorter: { ...tableState.sorter }, geneNamesFilter: searchPattern };
    dispatch(updateGeneList(experimentID, newTableState));
    setGeneNamesFilter(searchPattern);
  };

  return (
    <Space direction='vertical' style={{ width: '100%' }}>
      <FilterGenes filterGenes={filterGenes} />
      <Table
        columns={columns}
        dataSource={rows}
        loading={isLoading}
        size='small'
        pagination={tableState?.pagination}
        sorter={tableState?.sorter}
        onChange={handleTableChange}
      />
    </Space>
  );
};


GeneListTool.defaultProps = {};

GeneListTool.propTypes = {
  experimentID: PropTypes.string.isRequired,
};

export default GeneListTool;
