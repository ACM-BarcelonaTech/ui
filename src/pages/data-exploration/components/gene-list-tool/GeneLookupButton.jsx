
import React from 'react';
import { Tooltip, Button } from 'antd';
import { EyeTwoTone, EyeOutlined } from '@ant-design/icons';
import PropTypes from 'prop-types';

const GeneLookupButton = (props) => {
  const { focused, onClick } = props;

  return (
    <Tooltip placement='right' title='Show gene expression'>
      <Button type='text' size='small' onClick={onClick}>
        {focused
          ? (<EyeTwoTone style={{ cursor: 'pointer' }} />)
          : (<EyeOutlined style={{ cursor: 'pointer' }} />)}
      </Button>
    </Tooltip>
  );
};

GeneLookupButton.defaultProps = {
  onClick: () => null,
};

GeneLookupButton.propTypes = {
  focused: PropTypes.bool.isRequired,
  onClick: PropTypes.func,
};

export default GeneLookupButton;
