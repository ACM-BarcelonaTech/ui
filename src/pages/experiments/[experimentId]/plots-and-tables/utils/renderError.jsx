import React from 'react';
import {
  Button, Empty, Typography,
} from 'antd';
import { ExclamationCircleFilled } from '@ant-design/icons';

const { Text } = Typography;

const renderError = (err, onTryAgain) => (
  <Empty
    image={(
      <Text type='danger'>
        <ExclamationCircleFilled style={{ fontSize: 40 }} />
      </Text>
    )}
    imageStyle={{
      height: 40,
    }}
    description={
      err
    }
  >
    <Button
      type='primary'
      onClick={() => onTryAgain()}
    >
      Try again
    </Button>
  </Empty>
);

export default renderError;
