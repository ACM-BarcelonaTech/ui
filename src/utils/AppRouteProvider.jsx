import React, { useContext, useState } from 'react';
import propTypes from 'prop-types';
import { useRouter } from 'next/router';
import { useSelector } from 'react-redux';

import DataProcessingIntercept from 'components/data-processing/DataProcessingIntercept';

const AppRouterContext = React.createContext(null);

const AppRouteProvider = (props) => {
  const { children } = props;
  const router = useRouter();

  const [renderIntercept, setRenderIntercept] = useState(null);

  const changedQCFilters = useSelector(
    (state) => state.experimentSettings.processing.meta.changedQCFilters,
  );

  const availableIntercepts = {
    DATA_PROCESSING: (nextRoute, hardNavigate) => (
      <DataProcessingIntercept
        onContinueNavigation={() => continueNavigation(nextRoute, hardNavigate)}
        onDismissIntercept={() => setRenderIntercept(null)}
      />
    ),
  };

  const continueNavigation = (nextRoute, hardNavigate) => {
    // Hard navigate, causing the page to refresh and fetch data from server
    if (hardNavigate) window.location.href = nextRoute;
    router.push(nextRoute);
  };

  const handleRouteChange = (previousRoute, nextRoute, hardNavigate = false) => {
    if (previousRoute.match('/data-processing') && changedQCFilters.size > 0) {
      setRenderIntercept(availableIntercepts.DATA_PROCESSING(nextRoute, hardNavigate));
      return;
    }
    continueNavigation(nextRoute, hardNavigate);
  };

  const navigateTo = (
    nextRoute,
    refreshPage,
  ) => handleRouteChange(router.pathname, nextRoute, refreshPage);

  return (
    <AppRouterContext.Provider value={navigateTo}>
      {renderIntercept ?? <></>}
      {children}
    </AppRouterContext.Provider>
  );
};

AppRouteProvider.propTypes = {
  children: propTypes.node.isRequired,
};

const useAppRouter = () => useContext(AppRouterContext);

export { useAppRouter };
export default AppRouteProvider;
