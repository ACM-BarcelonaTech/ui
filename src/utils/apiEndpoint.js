const getApiEndpoint = () => {
  try {
    if (window.location.href.includes('biomage-ui-staging')) return 'https://biomage-api-staging.k8s-staging.biomage.net';
    if (window.location.href.includes('localhost')) return 'http://localhost:3000';
  } catch (error) {
    console.error('Failed to get API endpoint', window.location.href);
  }
  return 'https://biomage-api.k8s-staging.biomage.net';
};

export default getApiEndpoint;
