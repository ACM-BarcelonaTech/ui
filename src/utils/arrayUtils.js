import _ from 'lodash';

const isSubset = (
  (subsetArray, containingArray) => _.difference(subsetArray, containingArray).length === 0);

// eslint-disable-next-line import/prefer-default-export
export { isSubset };
