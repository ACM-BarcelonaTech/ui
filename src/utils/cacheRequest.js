/* eslint-disable no-underscore-dangle */
import crypto from 'crypto';
import hash from 'object-hash';
import cache from './cache';
import sendWork from './sendWork';
import isBrowser from './environment';
import CustomError from './customError';

const createHash = (endpoint) => crypto.createHash('md5').update(endpoint).digest('hex');
const createObjectHash = (object) => hash.MD5(object);

// eslint-disable-next-line no-unused-vars
const objectToSortedString = (object) => {
  let sortedString = '';
  Object.keys(object).sort().forEach((key) => {
    sortedString = `${sortedString}${key}${object[key]}`;
  });
  return sortedString;
};


const cacheFetch = async (endpoint, options = {}, ttl = 900) => {
  throw new Error('not currently implemented -- buggy, needs concurrency support');

  /* eslint-disable no-unreachable */
  if (isBrowser) {
    // We ask the cache for a result if we don't specify a method or it's GET.
    const isGet = !options.method || options.method === 'GET';

    // IMPORTANT: endpoint needs to be converted into an ordered list of query args
    // can use objectToSortedString for this with a URL parser.
    const key = createHash(`${endpoint}`);

    // Check if it's in the cache at all.
    if (isGet) {
      const data = await cache.get(key);
      if (data) {
        return data;
      }
    }

    // Get response.
    const response = await fetch(endpoint, options);

    // If not an OK code, throw error.
    if (!response.ok) {
      throw new Error(
        `Fetch to endpoint ${endpoint} with options ${options} failed, response code is ${response.status}`,
      );
    }

    // Convert to JSON.
    const json = await response.json();

    // If get, set the cache with the new result.
    // If not, reset the cache as the request has changed the values.
    if (isGet) {
      await cache._set(key, json, ttl);
    } else {
      try {
        await cache._remove(key);
      } catch (e) {
        // Removal didn't happen as key is not in cache. Not a problem.
      }
    }

    return json;
  }
  throw new Error('Disabling network interaction on server');
  /* eslint-enable no-unreachable */
};

const getFromApiExpectOK = async (url) => {
  const response = await fetch(url);

  if (response.ok) {
    const data = await response.json();
    return data;
  }

  throw new CustomError('There has been an error fetching the data.', response);
};

const decomposeBody = async (body, experimentId) => {
  const { genes: requestedGenes } = body;
  const missingDataKeys = {};
  const cachedData = {};

  await Promise.all(requestedGenes.map(async (g) => {
    const newBody = {
      ...body,
      genes: g,
    };
    const key = createObjectHash({ experimentId, newBody });
    const data = await cache.get(key);
    if (data) {
      cachedData[g] = data;
    } else {
      missingDataKeys[g] = key;
    }
  }));

  return { missingDataKeys, cachedData };
};

const fetchCachedGeneExpressionWork = async (experimentId, timeout, body, ttl = 900) => {
  const { missingDataKeys, cachedData } = await decomposeBody(body, experimentId);
  const missingGenes = Object.keys(missingDataKeys);
  if (missingGenes.length === 0) {
    return cachedData;
  }
  const response = await sendWork(experimentId, timeout, { ...body, genes: missingGenes });

  const responseData = JSON.parse(response.results[0].body);
  Object.keys(missingDataKeys).forEach(async (gene) => {
    await cache._set(missingDataKeys[gene], responseData[gene], ttl);
  });

  return responseData;
};

const fetchCachedWork = async (experimentId, timeout, body, ttl = 900) => {
  if (isBrowser) {
    if (body.name === 'GeneExpression') {
      return fetchCachedGeneExpressionWork(experimentId, timeout, body, ttl);
    }
    const key = createObjectHash({ experimentId, body });
    const data = await cache.get(key);
    if (data) return data;
    const response = await sendWork(experimentId, timeout, body);
    const responseData = JSON.parse(response.results[0].body);
    await cache._set(key, responseData, ttl);
    return responseData;
  }
  throw new Error('Disabling network interaction on server');
};


export {
  cacheFetch, getFromApiExpectOK, fetchCachedWork, fetchCachedGeneExpressionWork,
};
