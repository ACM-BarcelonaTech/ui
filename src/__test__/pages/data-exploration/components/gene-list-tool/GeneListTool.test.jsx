import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import preloadAll from 'jest-next-dynamic';
import thunk from 'redux-thunk';
import GeneListTool from '../../../../../pages/data-exploration/components/gene-list-tool/GeneListTool';
import GeneLookupButton from '../../../../../pages/data-exploration/components/gene-list-tool/GeneLookupButton';

import { fetchCachedWork } from '../../../../../utils/cacheRequest';

jest.mock('localforage');

jest.mock('../../../../../utils/cacheRequest', () => ({
  fetchCachedWork: jest.fn(() => new Promise((resolve) => resolve(JSON.stringify({
    rows: [{
      key: '1',
      gene_names: 'a mock name',
    }],
    total: 1,
  })))),
}));

const mockStore = configureMockStore([thunk]);
let component;
let store;

const initialState = {
  focusedGene: {},
  geneList: {
    loading: false,
    rows: [
      {
        gene_names: 'CEMIP',
        dispersions: 3.999991789324,
        key: 'CEMIP',
      },
      {
        gene_names: 'TIMP3',
        dispersions: 3.4388329,
        key: 'TIMP3',
      },
      {
        gene_names: 'SMYD3',
        dispersions: 3.1273264798,
        key: 'SMYD3',
      },
      {
        gene_names: 'CHI3L1',
        dispersions: 2.3248394823,
        key: 'CHI3L1',
      },
      {
        gene_names: 'NEAT1',
        dispersions: 3.3243243231,
        key: 'NEAT1',
      },
      {
        gene_names: 'GPC6',
        dispersions: 1.8934782,
        key: 'GPC6',
      },
      {
        gene_names: 'CEMIP',
        dispersions: 0.6854982354,
        key: 'CEMIP',
      },
      {
        gene_names: 'A',
        dispersions: 0.454705249,
        key: 'A',
      },
      {
        gene_names: 'B',
        dispersions: 1.4854934373,
        key: 'B',
      },
      {
        gene_names: 'C',
        dispersions: 2.12143434322,
        key: 'C',
      },
      {
        gene_names: 'D',
        dispersions: 0.8423643682427,
        key: 'D',
      },
      {
        gene_names: 'E',
        dispersions: 2.869048545643,
        key: 'E',
      },
      {
        gene_names: 'F',
        dispersions: 1.3456954368039,
        key: 'F',
      },
      {
        gene_names: 'G',
        dispersions: 0.5865059484,
        key: 'G',
      },
      {
        gene_names: 'H',
        dispersions: -0.4884802,
        key: 'H',
      },
      {
        gene_names: 'I',
        dispersions: 0.08756543,
        key: 'I',
      },
      {
        gene_names: 'J',
        dispersions: -0.12243687564,
        key: 'J',
      },
      {
        gene_names: 'K',
        dispersions: 0.46765344343,
        key: 'K',
      },
      {
        gene_names: 'L',
        dispersions: -0.545768743,
        key: 'L',
      },
    ],
    tableState: {
      pagination: {
        current: 1,
        pageSize: 15,
        showSizeChanger: true,
        total: 19,
      },
      sorter: {
        field: 'dispersions',
        order: 'descend',
      },
    },
  },
  selectedGenes: { geneList: {} },
  geneExpressiondata: { data: {} },
};

describe('GeneListTool', () => {
  beforeAll(async () => {
    await preloadAll();
  });

  beforeEach(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });

    store = mockStore(initialState);

    component = mount(
      <Provider store={store}>
        <GeneListTool experimentID='1234' />
      </Provider>,
    );
  });

  afterEach(() => {
    component.unmount();
  });

  configure({ adapter: new Adapter() });
  it('renders correctly', () => {
    const table = component.find('Table Table');
    const spin = component.find('Table Spin');
    const genesFilter = component.find('FilterGenes');
    expect(spin.length).toEqual(1);
    expect(table.length).toEqual(1);
    expect(genesFilter.length).toEqual(1);
    expect(table.getElement().props.columns.length).toEqual(3);
    expect(table.getElement().props.columns[0].title).toEqual('Gene');
    expect(table.getElement().props.columns[1].key).toEqual('lookup');
    expect(table.getElement().props.columns[2].title).toEqual('Dispersion');
    expect(table.getElement().props.dataSource.length).toEqual(19);
    expect(table.getElement().props.data.length).toEqual(15);
  });

  it('can sort the gene names in alphabetical order', () => {
    const newPagination = {
      current: 1,
      pageSize: 15,
      showSizeChanger: true,
      total: 19,
    };

    const newSorter = {
      column: {
        dataIndex: 'gene_names',
        key: 'gene_names',
      },
      render: jest.fn(),
      columnKey: 'gene_names',
      field: 'gene_names',
      order: 'ascend',
    };

    const table = component.find('Table Table');
    const _ = {};
    table.getElement().props.onChange(newPagination, _, newSorter);
    table.update();
    expect(fetchCachedWork).toHaveBeenCalledWith('1234', 30, {
      limit: 15, name: 'ListGenes', offset: 0, orderBy: 'gene_names', orderDirection: 'ASC', selectFields: ['gene_names', 'dispersions'],
    });
    expect(store.getActions().length).toEqual(1);
    expect(store.getActions()[0].type).toEqual('GENE_LIST.LOAD');
  });

  it('All `eye` buttons are initially unfocused.', () => {
    const table = component.find('Table Table');

    table.getElement().props.data.forEach((row) => {
      expect(row.lookup.props.focused).toEqual(false);
    });
  });

  it('Clicking one of the `eye` buttons triggers appropriate onChange actions.', () => {
    const table = component.find('Table Table');

    const { onClick } = table.getElement().props.data[2].lookup.props;

    // trigger clicking
    onClick();

    // The store should have been updated.
    expect(store.getActions().length).toEqual(1);
    expect(store.getActions()[0]).toMatchObject({ type: 'SET_FOCUSED_GENE' });
  });

  it('Having a focused gene triggers focused view for `eye` button.', () => {
    const FOCUSED_GENE = 'NEAT1';

    // Redefine store from `beforeEach`.
    store = mockStore({ ...initialState, focusedGene: { geneName: FOCUSED_GENE } });
    component = mount(
      <Provider store={store}>
        <GeneListTool experimentID='1234' />
      </Provider>,
    );

    const table = component.find('Table Table');

    table.getElement().props.data.forEach((row) => {
      if (row.gene_names === FOCUSED_GENE) {
        expect(row.lookup.props.focused).toEqual(true);
      } else {
        expect(row.lookup.props.focused).toEqual(false);
      }
    });
  });
});
