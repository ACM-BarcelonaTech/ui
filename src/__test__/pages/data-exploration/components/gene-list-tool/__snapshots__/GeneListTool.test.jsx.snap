// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneListTool Clicking one of the \`eye\` buttons triggers appropriate onChange actions. 1`] = `
Object {
  "data": Object {
    "isLoading": true,
  },
  "type": "SET_FOCUSED_GENE",
}
`;
